### ISSUE TITLE
clang-format complaing about unknown key after I deleted the bundled executable?

### ISSUE THREAD

**Issue by dariomx** [microsoft/vscode-cpptools#9910]
### Environment

- OS and Version: Windows 10
- VS Code Version: 1.71.1
- C/C++ Extension Version: 1.12.4
- Other extensions you installed (and if the issue persists after disabling them):  remote - ssh, clangd
- If using SSH remote, specify OS of remote machine:  Linux
- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc).

See bug summary below

### Bug Summary and Steps to Reproduce

Bug Summary:

When trying to format a .c file I see this error on clangd log:

/home/user/.clang-format:61:1: error: unknown key 'SeparateDefinitionBlocks'
SeparateDefinitionBlocks: Always

But if I run the bundled clang-format manually from .vscode-server/extensions/ms-vscode.cpptools-1.12.4-linux-x64/LLVM/bin, it is able to format the same file without errors (I double checked it was reading same settings from .clang-format).

The clang-format documentation for that tag says is supported on version 14, which is the one bundled. So I am puzzled ...

Just for the sake of it, I deleted the clang-format executable from the mentioned path, expecting to get a different error (missing executable or something). To my surprise though, it continues to complain about the unknown key! I thought that error was coming from clang-format itself, but per my experiment is coming from some other place?

Please clarify and advise how to fix this,
Thanks.

Steps to reproduce:

### Expected behavior

.

### Code sample and Logs

```shell
.
```

### Screenshots

.

**Comment by sean-mcmanus** [comment]
Do you have the C_Cpp.clang_format_path setting set? Does another clang-format exist in your path? The code we use to select the clang-format to use is at https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/settings.ts#L111 . We try to choose the newer version.

If you set setting
```json
    "C_Cpp.codeAnalysis.clangTidy.args": [
        "--version"
    ],
```
That will show the version of clang-tidy used, which probably should match the clang-format version used as well.

**Comment by sean-mcmanus** [comment]
Oh, it looks like you're using the clangd extension and not the C/C++ extension? It's probably using the formatter installed in your path.

**Comment by sean-mcmanus** [comment]
You may be hitting https://github.com/microsoft/vscode-cpptools/issues/7083 .

**Comment by dariomx** [comment]
Hello Sean,

Thanks for your quick reply. I am using both C/C++ and clangd extensions, and I am pretty sure that the only clang-format executable available on my remote box is the come bundled with this extension.

I would like to print the path of the clang-format being used. I tried to do this on the main.js of the extension with no results, could you assist in doing that? That should solve the mystery.

Thanks.

PS: Let us reopen this please, as the issue is not resolved yet.

**Comment by dariomx** [comment]
Hi Sean,

Thanks for the quick reply. I thought more about my complain and I think I misinterpreted the information. The fact that the unknown key came from clangd extension log, and not from C/C++ log, should have suggested that the clang-format bundled with this extension was not the issue. I think you were under this idea as well, when asked if I was using both extensions.

As you suggested as well, I have disabled intellisense for the sake of using clangd features. So that is further support to say that the formatting of this extension is not being used indeed (due https://github.com/microsoft/vscode-cpptools/issues/7083, as you pointed out on your last comment).

Based on this reflections I went back to my childhood ... wait, not that long. But I went back to find that clangd, which I am using, embeds clang-format (https://clangd.llvm.org/features). This means I am restricted to whatever clang-format version they used, when building the clangd version I have installed. But in that sense, I found this other issue quite useful:

https://github.com/clangd/vscode-clangd/issues/191

As it points out a very similar complain and suggests a workaround: to install extension xaver.clang-format,  to offer another provider of clang-format which overwrites the old version from llvm-vs-code-extensions.vscode-clangd. In order to accomplish that, I installed the mentioned extension and setup these two properties on my settings.json:

```
    "clang-format.executable": "/path/to/my/newer/clang-format",
    "[c]": {
        "editor.defaultFormatter": "xaver.clang-format"
    }
```

After that I no longer see the unknown key error due my newer clang-format version. I am happy now.

Thanks.
